pycobertura/filesystem.py

```diff
@@ -68,3 +71,106 @@ def open(self, filename):
+            prefix=prefix, ref=self.ref, filename=filename
+        )
+
+    def has_file(self, filename):
+        command = "git --no-pager show {}".format(self.real_filename(filename))
+        command_tokens = shlex.split(command)
+        return_code = subprocess.call(
+            command_tokens,
+            cwd=self.repository,
+            shell=True,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        return not bool(return_code)
+
+    def _get_root_path(self, repository_folder):
+        command = "git rev-parse --show-toplevel"
+        command_tokens = shlex.split(command)
+        try:
+            output = subprocess.check_output(
+                command_tokens, cwd=repository_folder
+            )
+        except (OSError, subprocess.CalledProcessError):
+            raise ValueError(
+                "The folder {} is not "
+                "a valid git repository.".format(repository_folder)
+            )
+
+        return output.decode("utf-8").rstrip()
+
+    @contextmanager
+    def open(self, filename):
+        """
+        Yield a file-like object for file `filename`.
+
+        This function is a context manager.
+        """
+        filename = self.real_filename(filename)
+
+        command = "git --no-pager show {}".format(filename)
+        command_tokens = shlex.split(command)
+
+        try:
+            output = subprocess.check_output(
+                command_tokens, cwd=self.repository
+            )
+        except (OSError, subprocess.CalledProcessError):
+            raise self.FileNotFound(filename)
+
+        output = output.decode("utf-8").rstrip()
+        yield io.StringIO(output)
+
+
+def filesystem_factory(report=None, source=None, source_prefix=None, ref=None):
+    """
+    The optional argument `report` is the location of the cobertura report.
+    It will be used to build the sources path.
+ 
+    The optional argument `source` is the location of the source code
+    provided as a directory path or a file object zip archive containing
+    the source code.
+
+    The optional argument `source_prefix` will be used to lookup source
+    files if a zip archive is provided and will be prepended to filenames
+    found in the coverage report.
+
+    The oprional argument `ref` will be taken into account when
+    instantiating a GitFileSystem, and it shall be a branch name, a commit
+    ID or a git ref ID.
+    """
+    if source is None:
+        if isinstance(report, str):
+            # get the directory in which the coverage file lives
+            source = os.path.dirname(report)
+
+    if zipfile.is_zipfile(source):
+        return ZipFileSystem(
+            source, source_prefix=source_prefix
+        )
+
+    if os.path.isfile(source):
+        source = os.path.dirname(source)
+
+    if ref:
+        return GitFileSystem(source, ref)
+    else:
+        return DirectoryFileSystem(source, source_prefix=source_prefix)
```

Comment:
We can avoid the else-statement here since we use a guard-clause above.

```suggestion

    return DirectoryFileSystem(source, source_prefix=source_prefix)
```